calculator/calculator.py

```diff
@@ -1,15 +1,123 @@
 import re
+import math
+from typing import Union
+from statistics import mean
 
 
-class Calculator:
+_NUMBER_REGEX = r'\d+(?:\.\d+)?'
+_NUMBER_NEGATIVE_OPT_REGEX = f'-?{_NUMBER_REGEX}'
+_LEFT_PARENTHESIS_REGEX = r'(?P<LEFT_PARENTHESIS>\()'
+_RIGHT_PARENTHESIS_REGEX = r'(?P<RIGHT_PARENTHESIS>\))'
+_FACTORIAL_REGEX = r'(?P<FACTORIAL>!)'
+_NEGATE_REGEX = r'(?P<NEGATE>~)'
+_AVERAGE_REGEX = r'(?P<AVERAGE>@)'
+_MIN_REGEX = r'(?P<MIN>\&)'
+_MAX_REGEX = r'(?P<MAX>\$)'
+_POWER_REGEX = r'(?P<POWER>\^)'
+_MULTIPLICATION_REGEX = r'(?P<MULTIPLICATION>\*)'
+_DIVISION_REGEX = r'(?P<DIVISION>/)'
+_ADDITION_REGEX = r'(?P<ADDITION>\+)'
+_SUBTRACTION_REGEX = r'(?P<SUBTRACTION>-)'
 
-    @classmethod
-    def calculate(cls, expression=str):
-        if type(expression) is not str:
-            raise TypeError("Expression input should be of str type.")
 
-        # TODO
+def calculate(expression: str) -> Union[int, float]:
+    if type(expression) is not str:
+        raise TypeError("Expression input should be of str type.")
 
+    return _calculate(expression.replace(" ", ""))
 
+def _calculate(expression: str) -> Union[int, float]:
+    if _is_end_of_calculation(expression):
+        return expression
 
+    for regex, handler in _TOKEN_SPECIFICATION.items():
+        tmp_result = _handle(expression, regex, handler)
+        if tmp_result is not None:
+            return _calculate(tmp_result)
```

Comment:
```suggestion
    for regex, handler in _OPERATIONS.items():
        match = re.search(regex, expression)
        if match:
            solved_step_expression = re.sub(token.re, handler(match), expression, 1)
            return _calculate(solved_step_expression)  
```
I think that _handle function is redundent. 
Also, give your variables full and indicative names, it's Python, I should read english.